File: CS/E3507/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraGrid.Selection;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Columns;
using E1271;

namespace E3507 {
    public partial class Form1 : Form {
        Dictionary<int, GridCheckMarksSelection> detailCache;
        public Form1() {
            InitializeComponent();
            ds = InitData();
            gridControl1.DataSource = ds.MasterTable;
            new GridCheckMarksSelection(gridView1);

        }
        DataSet1 InitData()
        {
            DataSet1 dataSet = new DataSet1();
            for (int i = 0; i < 5; i++)
            {

                dataSet.MasterTable.Rows.Add(i, "Master " + i);
                for (int j = 0; j < 5; j++)
                {
                    dataSet.Child2.Rows.Add(i * 100 + j, i, "Child2:" + j);
                }
            }
            return dataSet;

        }
        DataSet1 ds;
        private void Form1_Load(object sender, EventArgs e) {
            // TODO: This line of code loads data into the 'nwindDataSet.Orders' table. You can move, or remove it, as needed.
            this.gridView1.ExpandAllGroups();

        }
        private void gridView1_MasterRowExpanded(object sender, DevExpress.XtraGrid.Views.Grid.CustomMasterRowEventArgs e)
        {
            GridView view = sender as GridView;
            GridView detailView = view.GetDetailView(e.RowHandle, e.RelationIndex) as GridView;
            int SourceIndex = gridView1.GetDataSourceRowIndex(e.RowHandle);
            if (detailCache == null)
                detailCache = new Dictionary<int, GridCheckMarksSelection>();
            if (!detailCache.ContainsKey(SourceIndex))
            {
                GridCheckMarksSelection detailHelper = new GridCheckMarksSelection();
                detailHelper.DetailViewAttach(detailView);
                detailCache.Add(SourceIndex, detailHelper);
            }
            else
                detailCache[SourceIndex].DetailViewAttach(detailView);
        
        }

        private void gridView1_Click(object sender, EventArgs e)
        {

        }

        private void gridView1_MasterRowCollapsed(object sender, CustomMasterRowEventArgs e)
        {
        }

        private void gridView1_MasterRowExpanding(object sender, MasterRowCanExpandEventArgs e)
        {

        }

        private void gridView1_MasterRowCollapsing(object sender, MasterRowCanExpandEventArgs e)
        {
            GridView view = sender as GridView;
            GridView detailView = view.GetDetailView(e.RowHandle, e.RelationIndex) as GridView;
            int SourceIndex = gridView1.GetDataSourceRowIndex(e.RowHandle);
            if (detailCache == null)
                return;
            if (detailCache.ContainsKey(SourceIndex))
                detailCache[SourceIndex].DetailViewDetach();


        }
    }
}

File: VB/E3507/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraGrid.Selection
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Columns
Imports E1271

Namespace E3507
	Partial Public Class Form1
		Inherits Form
		Private detailCache As Dictionary(Of Integer, GridCheckMarksSelection)
		Public Sub New()
			InitializeComponent()
			ds = InitData()
			gridControl1.DataSource = ds.MasterTable
			Dim TempGridCheckMarksSelection As GridCheckMarksSelection = New GridCheckMarksSelection(gridView1)

		End Sub
		Private Function InitData() As DataSet1
			Dim dataSet As New DataSet1()
			For i As Integer = 0 To 4

				dataSet.MasterTable.Rows.Add(i, "Master " & i)
				For j As Integer = 0 To 4
					dataSet.Child2.Rows.Add(i * 100 + j, i, "Child2:" & j)
				Next j
			Next i
			Return dataSet

		End Function
		Private ds As DataSet1
		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			' TODO: This line of code loads data into the 'nwindDataSet.Orders' table. You can move, or remove it, as needed.
			Me.gridView1.ExpandAllGroups()

		End Sub
		Private Sub gridView1_MasterRowExpanded(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Grid.CustomMasterRowEventArgs) Handles gridView1.MasterRowExpanded
			Dim view As GridView = TryCast(sender, GridView)
			Dim detailView As GridView = TryCast(view.GetDetailView(e.RowHandle, e.RelationIndex), GridView)
			Dim SourceIndex As Integer = gridView1.GetDataSourceRowIndex(e.RowHandle)
			If detailCache Is Nothing Then
				detailCache = New Dictionary(Of Integer, GridCheckMarksSelection)()
			End If
			If (Not detailCache.ContainsKey(SourceIndex)) Then
				Dim detailHelper As New GridCheckMarksSelection()
				detailHelper.DetailViewAttach(detailView)
				detailCache.Add(SourceIndex, detailHelper)
			Else
				detailCache(SourceIndex).DetailViewAttach(detailView)
			End If

		End Sub

		Private Sub gridView1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles gridView1.Click

		End Sub

		Private Sub gridView1_MasterRowCollapsed(ByVal sender As Object, ByVal e As CustomMasterRowEventArgs) Handles gridView1.MasterRowCollapsed
		End Sub

		Private Sub gridView1_MasterRowExpanding(ByVal sender As Object, ByVal e As MasterRowCanExpandEventArgs) Handles gridView1.MasterRowExpanding

		End Sub

		Private Sub gridView1_MasterRowCollapsing(ByVal sender As Object, ByVal e As MasterRowCanExpandEventArgs) Handles gridView1.MasterRowCollapsing
			Dim view As GridView = TryCast(sender, GridView)
			Dim detailView As GridView = TryCast(view.GetDetailView(e.RowHandle, e.RelationIndex), GridView)
			Dim SourceIndex As Integer = gridView1.GetDataSourceRowIndex(e.RowHandle)
			If detailCache Is Nothing Then
				Return
			End If
			If detailCache.ContainsKey(SourceIndex) Then
				detailCache(SourceIndex).DetailViewDetach()
			End If


		End Sub
	End Class
End Namespace